### DISCUSSION TITLE
Difference between preview and run

### DISCUSSION THREAD

**Discussion by jchapuis** [VirtusLab/besom#453] [Looking for feedback]
Somehow I ended up with a non-deterministic situation where the preview is sometimes different than what happens when running, is this expected? trace below. 

(Note that I use for comprehensions and I don't list all outputs in the Stack constructor, I can paste the program if it helps, or try to find a simpler repro)

🙏 

```
View in Browser (Ctrl+O): https://app.pulumi.com/jchapuis/Example/hackintosh/previews/e7608c59-ae5b-4c6d-9231-fb8193a04572

     Type                               Name                                 Plan
 +   pulumi:pulumi:Stack                Example-hackintosh                   create
 +   ├─ aws:s3:Bucket                   s3-website-bucket                    create
 +   ├─ aws:s3:BucketPublicAccessBlock  s3-website-bucket-publicaccessblock  create
 +   ├─ aws:s3:BucketPolicy             s3-website-bucket-access-policy      create
 +   ├─ aws:s3:BucketObject             files.html                           create
 +   ├─ aws:s3:BucketObject             upload.html                          create
 +   ├─ aws:iam:User                    uploads                              create
 +   ├─ kubernetes:core/v1:Pod          troubleshooting                      create
 +   └─ aws:iam:AccessKey               exampleAccessKey                     create

Outputs:
    bucketName: "s3-website-bucket"
    websiteUrl: output<string>

Resources:
    + 9 to create

Do you want to perform this update? yes
Updating (hackintosh)

View in Browser (Ctrl+O): https://app.pulumi.com/jchapuis/Example/hackintosh/updates/1

     Type                                                    Name                                 Status
 +   pulumi:pulumi:Stack                                     Example-hackintosh                   created (29s)
 +   ├─ aws:s3:Bucket                                        s3-website-bucket                    created (0.62s)
 +   ├─ aws:s3:BucketPublicAccessBlock                       s3-website-bucket-publicaccessblock  created (0.57s)
 +   ├─ aws:s3:BucketPolicy                                  s3-website-bucket-access-policy      created (0.34s)
 +   ├─ aws:s3:BucketObject                                  files.html                           created (0.84s)
 +   ├─ aws:s3:BucketObject                                  upload.html                          created (1s)
 +   ├─ aws:iam:User                                         uploads                              created (0.25s)
 +   ├─ kubernetes:core/v1:Pod                               troubleshooting                      created (1s)
 +   ├─ aws:iam:AccessKey                                    exampleAccessKey                     created (0.24s)
 +   ├─ kubernetes:core/v1:ServiceAccount                    lease-access                         created (0.38s)
 +   ├─ kubernetes:rbac.authorization.k8s.io/v1:Role         lease-access                         created (0.25s)
 +   ├─ kubernetes:rbac.authorization.k8s.io/v1:RoleBinding  lease-access                         created (0.25s)
 +   └─ k8sx:service:ServiceDeployment                       uploads                              created (12s)
 +      ├─ kubernetes:apps/v1:Deployment                     uploads                              created (1s)
 +      ├─ kubernetes:core/v1:Service                        uploads                              created (10s)
 +      └─ kubernetes:networking.k8s.io/v1:Ingress           uploads                              created (0.77s)

Outputs:
    bucketName: "s3-website-bucket"
    websiteUrl: "s3-website-bucket.s3-website-us-east-1.amazonaws.com"

Resources:
    + 16 created

Duration: 38s
```

**Comment by lbialy** [comment]
Yes, this is, unfortunately, a part of the internal protocol that Besom has to implement like all other SDKs. The difference is that - being purely functional - Besom has to return resources wrapped in `Output`s from resource constructors. Other SDKs have sync constructors performing async side effects in said constructors but we didn't want or could to do that if we wanted to support CE and ZIO. This means however that we have two different semantics fused into one monad. When you call `val thingie = Output(1)` or `val another = Output.eval(catsEffectIO: IO[Thing])` or `config.requireBoolean("isCI"): Output[Boolean]` `Output` behaves just like `IO` would. When you access a property defined on a resource however things change. Resources are dynamic entites that do not exist until they are created and therefore it is possible for their values to be unknown. Outputs implement this condition and therefore have an `Option`'s semantics baked-in inside. When pulumi executes a Besom program in dry run mode most of these properties on resources get set internally to our equivalent of `None` and therefore short circuit and skip all the maps and flatMaps on them. This is fine if they are passed to another Resource's Args as part of configuration because Outputs carry metadata with them and therefore allow Pulumi to detect data dependencies between resources necessary to construct proper deployment graph that you see above as a plan from dry run. Pulumi is powerful and elastic enough to deal with the fact that plan can become dynamic if user needs it to be dynamic but in such cases what you see above happens - plan does not show everything that will happen in actual application of the infra because monads allow for dynamic subtrees. This problem is exacerbated in Scala, especially when coming from pure fp background, because we're so used to just use for-comps but for comprehensions are flatMaps and therefore make plan dynamic. This is the reason why all our [examples](https://github.com/VirtusLab/besom/blob/main/examples) have this flat structure and also why we disallowed returning `Output[Stack]` from the `Pulumi.run` main function. 

I have [improved our docs to better explain this logic](https://virtuslab.github.io/besom/docs/io/) but I feel that this is probably not clear enough. We are working on something that would allow us to detect when user does this and issue a compile-time warning about the dynamicity. There's also the open question whether we should have two types of Output monad or expose the internal Result monad to the public. Two types of Output could not only detect a call that incurs dynamicity for the deployment plan but also prevent mistakes by clearly separating local/static outputs from remote/computed property outputs and having flatMap implementations that properly propagate the infective nature of remote/computed property outputs. This however may make things more complex. We are open for suggestions, it's not to late to introduce such a change if it improves the usability.

**Reply by jchapuis** [reply]
Thanks for the answer! All clear I think, and the addition to the docs also helps. I grasp the challenge now. I'm not familiar at all with pulumi internals, but it feels that the culprit is a certain weakness from the dry run feature? In an ideal world, the dry-run should be able to simulate dynamicity as well, maybe with mocked APIs or some concept for typical return values for certain calls (that providers would have to fill-in). Intuitively I would say complicating besom's design due to a pulumi weakness would be sad. 
Is there a way that besom could take more control of the dry-run and somehow tell pulumi more about dynamic dependencies? At some extreme, maybe besom could somehow decorate the graph or provide its own dry-run representation 😄

**Reply by lbialy** [reply]
Yeah, this is only a problem in dry run, actual application of the infrastructure never uses `OutputData.unknown` (the `None` case). The problem with simulation is that when you have an `Output[SomeStruct]` and you do `output.flatMap { struct => doThingsWithStruct(struct) }` your code will do arbitrary things with the instance of `SomeStruct`, it may take decisions based on one or more values for instance. Let's assume we provided a `Default[A]` typeclass instances for any type that arrives over the wire from the engine and instead of using `OutputData.unknown` just use the default value. This would be most probably prohibitively expensive (a lot of code to generate, packages would get even bigger) for one but also could lead to plans showing completely different things than what actual run would do. Easiest example: `Default[Boolean]` and an `if` based on it's value. What we get from the wire is an information that resource properties are unknown. What we do with this on our side is our business but I feel we shouldn't introduce even more confusing scenarios. I'll post a draft of an experimental solution we've come up with in next comment here.

**Comment by lbialy** [comment]
Here's a small prototype of compile-time warning about resource constructor being used in `Output#flatMap` scope. It has some weaknesses, unfortunately, as depicted below. It does work and can be added as-is. I have no idea on how to deal with the opaqueness of arbitrary functions that blocks this scope. Maaaybe, just maybe, we could piggy-back this on a `Context` subtype?

```scala
//> using scala 3.3.3
//> using dep com.lihaoyi::sourcecode::0.3.1

package besom.experimental

import scala.compiletime.summonFrom
import sourcecode.*

trait Warn
case object Warn extends Warn

sealed trait CheckWarn:
  def warnIfNecessary(using Enclosing, Line): Unit =
    this match
      case LetsWarn => LetsWarn.warn
      case NoWarn   =>

case object LetsWarn extends CheckWarn:
  def warn(using Enclosing, Line): Unit = println(s"warned at runtime: ${summon[Enclosing].value}:${summon[Line].value}")

case object NoWarn extends CheckWarn

object CheckWarn:
  inline given CheckWarn = summonFrom:
    case _: Warn =>
      Macro.warning("You'd better not do this")
      LetsWarn
    case _ =>
      NoWarn

case class Output[+A](run: () => A):
  given Warn = Warn
  def map[B](f: A => B): Output[B] = Output(() => f(run()))
  def flatMap[B](f: Warn ?=> A => Output[B]): Output[B] = Output(() => f(run()).run())

object Output:
  def apply[A](a: A): Output[A] = Output(() => a)

object aws:
  object s3:
    case class Bucket(str: String)

    // this resource constructor detects whether it's being called in an Output#flatMap scope
    def Bucket(something: Int, somethingElse: String)(using CheckWarn, Enclosing, Line): Output[Bucket] =
      summon[CheckWarn].warnIfNecessary

      Output(Bucket(s"$something $somethingElse"))

// this function does not do detection
def normalFunction(something: Int, somethingElse: String): Output[String] =
  Output(s"$something $somethingElse")

// this function obfuscates the resource constructor, hides the scope information from it
// and therefore does not trigger the warning
def obfuscatedResourceConstructor(something: Int, somethingElse: String): Output[aws.s3.Bucket] =
  aws.s3.Bucket(something, somethingElse)

@main def main =

  val resOutput = aws.s3.Bucket(1, "fgsfds")

  val out = for
    x <- Output(1)
    y <- Output("fgsfds")
    res <- aws.s3.Bucket(x, y).flatMap(b => Output(b.str)) // proves no given collision
    res2 <- normalFunction(x, y)
    ohNo <- obfuscatedResourceConstructor(x, y)
  yield
    println(res)
    println(res2)

  resOutput.run()
  out.run()
```

```scala
package besom.experimental

import scala.quoted.*

object Macro:

  inline def warning(inline message: String): Unit = ${ warningImpl('message) }

  def warningImpl(message: Expr[String])(using Quotes): Expr[Unit] =
    val msg = message.valueOrAbort
    quotes.reflect.report.warning(msg)
    '{}
```

<img width="724" alt="Screenshot 2024-04-12 at 09 48 31" src="https://github.com/VirtusLab/besom/assets/8607442/4ebf0c94-ca93-4a9d-9f51-316d48305a41">

**Reply by lbialy** [reply]
ah, no, `Context` subtype would only allow runtime warning, compile-time is out because user will typically do `(using Context)` and not a specific `(using Context.FM)` (where `Context.FM` is this specific variant that would replace `Warn` in code above).

maybe that's good enough?

**Reply by jchapuis** [reply]
looks cool. Maybe such warnings coud be easily turned off with a different `Pulumi.run` overload, or some toggle, since in the end this is just about the dry-run, correct? Maybe for some programs and in some contexts, expressing everything with for comprehensions but losing the dry-run precision is acceptable.

**Comment by lbialy** [comment]
I have a feeling this is now possible with `tracked` - https://www.youtube.com/watch?v=06YvDUgeEQk @prolativ

**Reply by lbialy** [reply]
In fact, why don't we ask the author of `tracked` himself - @KacperFKorban - do you think we could use `tracked` to keep track of which `Output` is dynamic (is available only when actually deploying stuff to cloud, eg.: IP address of a new VPS) vs "static" (it is also possibly dynamic, eg.: read from configuration - but known in the dry run phase, doesn't require interaction with remote APIs). I think we could use tracked to differentiate `Output`s and finally solve this long-standing problem with a nice compile warning that can be `@nowarn`ed if user knows what the end result is going to be and is fine with it.

**Reply by jchapuis** [reply]
yep @lbialy good idea! 💡

**Reply by KacperFKorban** [reply]
We discussed it in DMs, but just to sum it up:
`tracked` will most likely **not** work here, since the user can always lose the additional type information by an explicit ascription or a method call without an explicitly precise return type.
IMHO a type (Singleton) parameter would work, but would worsen the DX.

**Reply by lbialy** [reply]
Yes, I had a misguided hope that `tracked` is able to somehow smuggle a hidden property (a type member for example) in a type that is otherwise a plain monad (`Output[A]`). The only way to make it work is to have a mechanism to propagate this property statically and to be able to propagate it statically, we'd have to have something like `R` type set in ZIO where all user code would have to keep it as a generic type to allow propagation. I've went back to the PoCs made by @prolativ in August of 2024 where he has tested out a lot of this and he has also proven that we'd need capture calculus as `flatMap` would have to provide a contextual capability demanded by all resource constructors that would then allow us to warn if that resource constructor is being called in the wrong kind of flatMap. All of that would break a lot of DX. 

Basically we'd have `type Output[+A, K]` where `K` is a type set which would be `Any` for local outputs (any outputs that have values known in dry run), `Remote` for outputs that have values known only in actual apply phase after dry run and `Resource` for resource constructors. Then in a situation where we'd have:
```scala
val out: Output[Something, Remote] = ???
out.flatMap { smth => 
  val res: Output[Bucket, Resource] = aws.s3.Bucket(... /* uses smth */)
  res // Warning: you are creating resources in a dynamic branch of code. These resources won't be visible in dry run and your infra will be different from what is planned!
}: Output[Bucket, Remote & Resource]
```
sort of, this is a very large change and has a lot of very weird corner cases.

**Reply by lbialy** [reply]
We had discussed this also with `Output[+A]` and `Output.Local[+A] <:< Output[+A]` too, instead of type param and the results were very similar.

**Reply by lbialy** [reply]
With a generic type set we possibly could drop the contextual capability altogether since we'd have aggregation of what was used the same way ZIO and Kyo has it. We have a lazy monadic api so things have to be composed together in the end and something that is not composed up into the returned value is a piece of dead code, fyi @prolativ.
